lib/diff.ml

```diff
@@ -101,13 +114,50 @@ and module_item ~typing_env ~name ~reference ~current =
       module_declaration ~typing_env ~name ~reference ~current
   | _ -> assert false
 
+and module_type_item ~typing_env ~name ~reference ~current =
+  match (reference, current) with
+  | None, None -> None
+  | None, Some (Modtype curr_mtd) ->
+      Some (Modtype { mtname = name; mtdiff = Added curr_mtd })
+  | Some (Modtype ref_mtd), None ->
+      Some (Modtype { mtname = name; mtdiff = Removed ref_mtd })
+  | Some (Modtype ref_mtd), Some (Modtype curr_mtd) ->
+      modtype_declaration ~typing_env ~name ~reference:ref_mtd ~current:curr_mtd
+  | _ -> assert false
+
 and module_declaration ~typing_env ~name ~reference ~current =
-  match (reference.md_type, current.md_type) with
+  module_type ~typing_env ~name ~ref_module_type:reference.md_type
+    ~current_module_type:current.md_type ~reference_location:reference.md_loc
+
+and modtype_declaration ~typing_env ~name ~reference ~current =
+  match (reference.mtd_type, current.mtd_type) with
+  | Some ref_sub, Some curr_sub ->
+      module_type ~typing_env ~name ~ref_module_type:ref_sub
+        ~current_module_type:curr_sub ~reference_location:reference.mtd_loc
+  | Some _ref_sub, None ->
```

Comment:
@panglesd 
Is this moving in the right direction?

Reply:
It is hard to say where it is moving, but you are close to the solution I'd be happy in this PR! Which is a solution which does not modify the types that you have defined.

Recall that `Some ...` means that the module type is not abstract: `module type S = sig ... end`.
`None` means that the module type is abstract: `module type S`.

So, when we are in the case where `current, reference` is `None, Some _` it means that the we went from an abstract module type to one with a signature. Does that best correspond to an addition removal or modification of a module type?
If it is an addition, what is added?
If it is a removal, what is removed?
If it is a modification, you have two ways to describe the modification: either by giving the list of modification (the `Supported` constructor, or by saying that the change is currently unsupported (the `Unsupported` constructor. Which is applicable here?

Try to think "high level", and only do the implementation when you know what you want to represent. Note that it is perfectly fine to say that a modification is `Unsupported`, supporting new kinds of modification is definitely for another PR!